### DISCUSSION TITLE
Implement alternative state-space dependence structure

### DISCUSSION THREAD

**Discussion by zcaiElvis** [kingaa/pomp#170] [Q&A] [answered]
First of all, thanks for making and maintaining this awesome package! 

Currently I am trying to build a particle filter of a state-space model, where the observations $y_i$ depends on both $x_i$, and all previous $y_1, ..., y_{i-1}$. I understand that pomp is more for a model where $y_i$ depends only on $x_i$, but I still wonder if this alternative dependence relationship is doable in pomp. 

Thanks for any help in advance!

**Comment by kingaa** [comment, marked as answer]
Good question, @zcaiElvis!

You are correct that the standard case that **pomp** treats is one for which, conditional on $x_i$, $y_i$ is independent of $y_j$ for all $j\ne{i}$.   However, as you are no doubt aware, the correctness of the particle filter does not depend on this, and it is possible to work around this restriction. 

The easiest way to do so is to include the data to the pomp object as "userdata", then to write the rmeasure and dmeasure components to make reference to this. This is really easy if you write rmeasure and dmeasure as **R** functions.  It is slightly more complex to write them as C snippets.

Consider the following very simple example. Suppose my data consist of cumulative counts and for some reason I do not want to simply take first differences. In particular, for each $i$, $y_i - y_{i-1} \sim \mathrm{Exponential}(x_i)$, where $x_i$ is the value of some state variable at time $t_i$.  Assuming that the data are stored in an **R** vector `y` and the observation times in a vector `t`, I could write something like
```
pfilter(
  Np = 1000,
  times = t, data = y,
  rprocess = <some rprocess code>,
  dmeasure = function (t, x, T, Y, ..., log) {
    i <- min(T>=t)
    dexp(x=Y[i]-Y[i-1], rate=x, log=log)
  },
  T = t, Y = y
)
```
The first line in the `dmeasure` code is not terribly elegant, but it should get the job done.

Now, the correctness of the particle filter *does* depend on the data not being anticipatory and you could easily get yourself into trouble making the $i$-th likelihood depend on data from the future.  Once you start going down this road of making the basic model components depend on userdata, it's up to you to ensure that your calculations are correct!

**Reply by zcaiElvis** [reply]
Thanks for the prompt reply, this is exactly what I am looking for!
